173 Cage Rework

TheDarkElites opened the pull request:
## About the Pull Request
173 has new cage mechanics! Wooh!
173 now has to damage its cage five times before escaping.
It can only damage its cage when not being watched.
The cage now needs to be repaired after being broken by 173.

## Why It's Good For The Game
Usually, when something that can only move when not being watched is in a cage the best way to contain it is to look at it, not move the cage around a bunch.

Also fixes #835

## Changelog

:cl:
add: Changed 173 cage mechanics
imageadd: New cage images so 173 can actually fit.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

<!--
Examples were changelog entries are optional/not typically required but encouraged:
* Cosmetic changes such as descriptions, sound effects, etc.
* Optimizations and other changes to underlying systems which do not affect gameplay.
* Minor bug fixes.

As a courtesy, for ported PRs, please include if you have the blessing of the author. While not required, we encourage basic decency in porting others' work. It is also sufficient to note that the author has not expressed displeasure at the idea of their work getting ported.
Please refrain from porting works of authors that have expressed displeasure in having their work ported, thank you.
-->

TheDarkElites committed:
Cage basics

TheDarkElites committed:
Merge branch 'dev' into 173-cage-rework

TheDarkElites commented:
Put this up early cause I had to get updated dev-branch so sshhhh don't look at anything

MACIEKBAKI commented on code/modules/SCP/SCPs/SCP-173.dm:
```diff
@@ -91,7 +91,19 @@ GLOBAL_LIST_EMPTY(scp173s)
 
 /mob/living/scp_173/UnarmedAttack(atom/A)
-	if(IsBeingWatched() || incapacitated()) // We can't do anything while being watched
+	if(incapacitated()) // We can't do anything while caged
+		if(!IsBeingWatched() && istype(A, /obj/structure/scp173_cage) && loc == "SCP-173 Containment Cage") //Can only attack the cage when not being watched and inside the cage
+			if(snap_cooldown > world.time)
+				var/obj/structure/scp173_cage/cage = A
+				cage.attack_hand(src)
+				snap_cooldown = world.time + snap_cooldown_time
+			else
+				to_chat(src, "<span class='warning'>You can't attack yet.</span>")
+		else if(IsBeingWatched() && istype(A, /obj/structure/scp173_cage) && loc == "SCP-173 Containment Cage")
+			to_chat(src, "<span class='warning'>You can't attack the cage while they're looking!</span>")
+		else
+			return
```
```suggestion
	if(incapacitated()) // We can't do anything while caged
		if(istype(A, /obj/structure/scp173_cage) && loc == A)
			var/obj/structure/scp173_cage/cage = A
			if(!IsBeingWatched()) //Can only attack the cage when not being watched and inside the cage
				if(snap_cooldown > world.time)
					cage.attack_hand(src)
					snap_cooldown = world.time + snap_cooldown_time
				else
					to_chat(src, SPAN_WARNING("You can't attack yet."))
					return
			else
				to_chat(src, SPAN_WARNING("You can't attack the cage while they're looking!"))
				return
		else
			return
	if(IsBeingWatched())// We can't do anything while being watched
```

MACIEKBAKI commented on code/modules/SCP/SCPs/SCP-173.dm:
```diff
@@ -343,14 +355,23 @@ GLOBAL_LIST_EMPTY(scp173s)
 	icon = 'icons/SCP/cage.dmi'
-	icon_state = "1"
-	name = "Empty SCP-173 Cage"
+	icon_state = "2"
```
what are these icon state names, who named them, jeez (not issue with pr, issue with whoever named the sprites)

MACIEKBAKI commented on code/modules/SCP/SCPs/SCP-173.dm:
```diff
@@ -359,30 +380,38 @@ GLOBAL_LIST_EMPTY(scp173s)
+		if(do_after(H, 5 SECONDS, src))
+			visible_message(SPAN_DANGER("[H] opens \the [src]!"))
+			ReleaseContents()
+	else if(istype(A, /mob/living/scp_173))
+		var/mob/living/scp_173/S = A
 		ReleaseContents()
```
hahahaha 173s can release other 173s 
tho why is `var/mob/living/scp_173/S = A` needed?
also I'd add a check to make sure the 173 that is releasing is *not* being watched / not caged itself

TheDarkElites replied:
that isn't done its just for testing purposes, currently working on it

TheDarkElites commented:
I would just like to once again point to my PR and say this isn't done and WIP. I had to put it up to marge dev branch with mine, so the first commit will look weird.

MACIEKBAKI commented:
((( i did not notice the fact its a draft )))

TheDarkElites commented:
> ((( i did not notice the fact its a draft )))

:)

TheDarkElites committed:
Cage Basics pt.2

- 173 Now attacks his own cage to damage
it
- Checks to make sure 173 is in the cage and not being watched

TheDarkElites committed:
Cage Logic

- New Cage Icons
- 173 now damages the cage through different stages
- Fixed HUD not loading while in cage
- Added ability to repair cage
- Added changing description based on cage damage value

TheDarkElites committed:
Removed debug stuff I accidently left in

aidan-bowman commented on code/modules/SCP/SCPs/SCP-173.dm:
```diff
@@ -91,7 +96,20 @@ GLOBAL_LIST_EMPTY(scp173s)
+				var/obj/structure/scp173_cage/cage = A
+				cage.attack_hand(src)
+				snap_cooldown = world.time + snap_cooldown_time
+				to_chat(src, "<span class='warning'>You damage the cage.</span>")
```
Use SPAN_WARNING and the other defines

TheDarkElites committed:
Switched stuff out with Macros

TheDarkElites committed:
Fixes previous commit

TheDarkElites committed:
Merge branch 'dev' into 173-cage-rework

TheDarkElites committed:
Adds more descriptive comments for cage procs and overrides

TheDarkElites commented:
@cheesePizza2 do you have any insight on why 173's image might be getting loaded on top of the cage instead of under it? Cause I'm kind of stuck. If you have some free time please take a look, I'd take any advice.

aidan-bowman commented:
> @cheesePizza2 do you have any insight on why 173's image might be getting loaded on top of the cage instead of under it? Cause I'm kind of stuck. If you have some free time please take a look, I'd take any advice.

I'll take a look. Without actually looking at the code, my best guess is either
- the cage is layered under 173 and you're just putting both sprites in the same area, OR
- the cage has a special sprite for when 173 is in the cage, and 173 just doesn't show at all (like a locker)

TheDarkElites commented:
>the cage is layered under 173 and you're just putting both sprites in the same area, OR
the cage has a special sprite for when 173 is in the cage, and 173 just doesn't show at all (like a locker)

Unfortunately, after a long time of testing, it's neither. Layers are appropriately separated, and there is no special sprite. My best guess is the overlay/underlay system doesn't take layering into account.

Tsurupeta commented:
> @cheesePizza2 do you have any insight on why 173's image might be getting loaded on top of the cage instead of under it? Cause I'm kind of stuck. If you have some free time please take a look, I'd take any advice.

Planes. Mob plane is higher than object plane.

TheDarkElites commented:
> Planes. Mob plane is higher than object plane.

Would changing the icon layer not fix this? Or are layers relative to planes?

Tsurupeta commented:
> > Planes. Mob plane is higher than object plane.
> 
> Would changing the icon layer not fix this? Or are layers relative to planes?

Second.

TheDarkElites commented:
In that case, im just gonna re-sprite the cage to contain 173. It would look better that way anyhow.

aidan-bowman commented:
Found the issue. Moving 173 into the cage didn't actually call the relevant code, since it all operates on the stored value, which is left unset.

aidan-bowman commented:
> > @cheesePizza2 do you have any insight on why 173's image might be getting loaded on top of the cage instead of under it? Cause I'm kind of stuck. If you have some free time please take a look, I'd take any advice.
> 
> Planes. Mob plane is higher than object plane.

Underlayer is lower than mob and object, plus cage is set to mob_layer + 0.05

aidan-bowman commented:
> In that case, im just gonna re-sprite the cage to contain 173. It would look better that way anyhow.

Probably not the greatest solution, makes it harder to maintain 173's sprite and also doesn't allow for him to have multiple icon states

TheDarkElites commented:
>Probably not the greatest solution, makes it harder to maintain 173's sprite and also doesn't allow for him to have multiple icon states

Greatest? Probably not, however literally the only solution as of right now. I've spent all morning changing the layers around as you are suggesting right now to get it to work. Fundamentally it just doesn't.

Also, as far as the icon states, if you're referring to facing various directions, then don't worry im spriting that in too.

This is absolutely not the best idea, but there's no real alternative (unless yall have ideas), and since 173 already got his resprite I severely doubt it's gonna get changed.

Ill leave in a TODO for a proper implementation later and leave in the clean empty cages in the cage.dmi to make it easier for whoever reworks it.

aidan-bowman commented:
> > Probably not the greatest solution, makes it harder to maintain 173's sprite and also doesn't allow for him to have multiple icon states
> 
> Greatest? Probably not, however literally the only solution as of right now. I've spent all morning changing the layers around as you are suggesting right now to get it to work. Fundamentally it just doesn't.
> 
> Also, as far as the icon states, if you're referring to facing various directions, then don't worry im spriting that in too.
> 
> This is absolutely not the best idea, but there's no real alternative (unless yall have ideas), and since 173 already got his resprite I severely doubt it's gonna get changed.
> 
> Ill leave in a TODO for a proper implementation later and leave in the clean empty cages in the cage.dmi to make it easier for whoever reworks it.

Lmao it's not working because you don't actually apply to the function to 173, you need to do S = dropping in the function that moves 173 into the cage because right now the function isn't actually working. Check the review changes I posted.

TheDarkElites commented:
>Lmao it's not working because you don't actually apply to the function to 173, you need to do S = dropping in the function that moves 173 into the cage because right now the function isn't actually working. Check the review changes I posted.

I don't push stuff till I'm done with it..... The code you see for anything in cage rn isn't done.

For example, I already have S set to the dropping it just isn't reflected in the most recent push because _I haven't pushed it yet_

aidan-bowman commented:
> > Lmao it's not working because you don't actually apply to the function to 173, you need to do S = dropping in the function that moves 173 into the cage because right now the function isn't actually working. Check the review changes I posted.
> 
> I don't push stuff till I'm done with it..... The code you see for anything in cage rn isn't done.
> 
> For example, I already have S set to the dropping it just isn't reflected in the most recent push because _I haven't pushed it yet_

Gotya

aidan-bowman commented:
Perhaps setting the icon resets the underlayer?

TheDarkElites committed:
New Cage Sprites

173 Now Directly spirited in so it looks right
Also some other stuff

TheDarkElites commented:
> Perhaps setting the icon resets the underlayer?

Even if it did it wouldn't matter as I set the underlayer after setting the icon, plus if that was the case 173 wouldn't show up in the cell at all during testing

aidan-bowman commented:
With the way the code works right now, you could probably have the cage inherit from lockers instead of structures.

TheDarkElites committed:
Fixes 173 icon issues

thank you Saint. Tsurupeta

TheDarkElites commented:
You can change the plane of something! I never thought of that!

TheDarkElites committed:
Merge branch 'dev' into 173-cage-rework

aidan-bowman commented:
Not much of a reason to keep depreciated sprites, it's not like we can't just access older versions of files.

TheDarkElites committed:
Cleanup

TheDarkElites committed:
removed deprecated cage sprites

TheDarkElites committed:
Minor fix

Tsurupeta commented on code/modules/SCP/SCPs/SCP-173.dm:
```diff
@@ -343,55 +361,131 @@ GLOBAL_LIST_EMPTY(scp173s)
+
+
+/obj/structure/scp173_cage/MouseDrop_T(atom/movable/dropping, mob/user) // When someone drags 173 onto the cage
+	if(damage_state < damage_max)
```
early return here

Tsurupeta commented on code/modules/SCP/SCPs/SCP-173.dm:
```diff
@@ -343,55 +361,131 @@ GLOBAL_LIST_EMPTY(scp173s)
+	// Damage state of cage
+	var/damage_state = 0
+	// Currently held 173
+	var/mob/living/scp_173/S
```
Don't use single fucking character, especially uppercase, for stored SCP.

TheDarkElites committed:
Cleanup pt.2

- Fixed weird use of nested ifs
- changed var name to something better
- BUGFIX: humans cant pull 173 anymore. (Had nothing to do with cages nor my pr but I noticed it and figured I should fix.)

TheDarkElites committed:
Merge branch 'dev' into 173-cage-rework

TheDarkElites committed:
Fixes me being an idiot and stupid

hopefully

Tsurupeta commented:
I don't see you nulling variables of cage and 173 in their Destroy

TheDarkElites committed:
Merge branch 'dev' into 173-cage-rework

TheDarkElites commented:
fixed conflicts, now working on actual review fixes

TheDarkElites committed:
Modified destroy procs to avoid lurking refrences

TheDarkElites committed:
Merge branch '173-cage-rework' of https://github.com/TheDarkElites/Foundation-19 into 173-cage-rework

TheDarkElites committed:
fixed indentation

- indentation
- grammar fix
